Terminal.Gui/Views/TreeView.cs

```diff
@@ -0,0 +1,1836 @@
+		/// <param name="model"></param>
+		/// <returns></returns>
+		public override IEnumerable<T> GetChildren (T model)
+		{
+			return childGetter.Invoke(model);
+		}
+	}
+
+	/// <summary>
+	/// Interface for all non generic members of <see cref="TreeView{T}"/>
+	/// </summary>
+	public interface ITreeView {
+		
+		
+		/// <summary>
+		/// Contains options for changing how the tree is rendered
+		/// </summary>
+		TreeStyle Style{get;set;}
+
+		/// <summary>
+		/// Removes all objects from the tree and clears selection
+		/// </summary>
+		void ClearObjects ();
+		
+		/// <summary>
+		/// Sets a flag indicating this view needs to be redisplayed because its state has changed.
+		/// </summary>
+		void SetNeedsDisplay ();
+	}
+
+	/// <summary>
+	/// Convenience implementation of generic <see cref="TreeView{T}"/> for any tree were all nodes implement <see cref="ITreeNode"/>
+	/// </summary>
+	public class TreeView : TreeView<ITreeNode> {
+
+		/// <summary>
+		/// Creates a new instance of the tree control with absolute positioning and initialises <see cref="TreeBuilder{T}"/> with default <see cref="ITreeNode"/> based builder
+		/// </summary>
+		public TreeView ()
+		{
+			TreeBuilder = new TreeNodeBuilder();
+			AspectGetter = o=>o == null ? "Null" : (o.Text ?? o?.ToString() ?? "Unamed Node");
+		}
+	}
+	
+	/// <summary>
+	/// Defines rendering options that affect how the tree is displayed
+	/// </summary>
+	public class TreeStyle {
+
+		/// <summary>
+		/// True to render vertical lines under expanded nodes to show which node belongs to which parent.  False to use only whitespace
+		/// </summary>
+		/// <value></value>
+		public bool ShowBranchLines {get;set;} = true;
+				
+		/// <summary>
+		/// Symbol to use for branch nodes that can be expanded to indicate this to the user.  Defaults to '+'. Set to null to hide
+		/// </summary>
+		public Rune? ExpandableSymbol {get;set;} = '+';
+				
+		/// <summary>
+		/// Symbol to use for branch nodes that can be collapsed (are currently expanded).  Defaults to '-'.  Set to null to hide
+		/// </summary>
+		public Rune? CollapseableSymbol {get;set;} = '-';
+
+		/// <summary>
+		/// Set to true to highlight expand/collapse symbols in hot key color
+		/// </summary>
+		public bool ColorExpandSymbol {get;set;}
+
+		/// <summary>
+		/// Invert console colours used to render the expand symbol
+		/// </summary>
+		public bool InvertExpandSymbolColors {get;set;}
+
+		/// <summary>
+		/// True to leave the last row of the control free for overwritting (e.g. by a scrollbar). When True scrolling will be triggered on the second last row of the control rather than the last.
+		/// </summary>
+		/// <value></value>
+		public bool LeaveLastRow {get;set;}
+
+	}
+	
+	/// <summary>
+	/// Hierarchical tree view with expandable branches.  Branch objects are dynamically determined when expanded using a user defined <see cref="ITreeBuilder{T}"/>
+	/// </summary>
+	public class TreeView<T> : View, ITreeView where T:class
```

Comment:
Why `where T: class` and not `where T: TreeNode`?  I am not 100% up to speed on generic type constraints yet, but it seems that you are missing an opportunity to ensure type safety. 


What am I missing?

Reply:
Thanks for diving into this so deeply!

The generic `TreeView<T>` supports any Type (e.g. `FileSystemInfo`) but requires that you specify an `ITreeBuilder` that shows how the objects relate to one another (how to determine children).

The non generic `TreeView` supports only `ITreeNode` objects and has a default `ITreeBuilder` already set up on it (`TreeNodeBuilder`).

Internally the TreeView stores the state of the tree as `Branch<T>`.  At the moment this is a private class and not visible to the API user.

The purpose of the regular `TreeView` is to mimic as closely as possible the WinForms tree view (create all nodes up front and then add them to the tree).

The purpose of the generic `TreeView<T>` is two fold
- To support hosting Typed objects e.g. if all your tree objects are `FileSystemInfo` then that is the Type that arrives in events (without having to change your class definition e.g. to implement an interface)
- To support late loading of children (e.g. until the user clicks expand on a node).  This is essential when fetching children may be relatively expensive e.g. listing database tables/columns or contacting a REST API.

When children can change over time (e.g. on a file system) the tree view can update itself to get a new authorative answer from the `ITreeBuilder` through methods like:

- RebuildTree
- RefreshObject

I have created an article (included in the PR) that describes from a user perspective how to create either type of tree:

docfx/articles/treeview.md